Update Get Started tutorial Part 1 with the Astronaut DAG (#3462)
astronomer/docs: learn/get-started-with-airflow.md

## Prerequisites

> - The [Astro CLI](https://docs.astronomer.io/astro/cli/install-cli).

Reviewer: We should probably say a version or greater here, since previous versions won't contain the astronauts example DAG.

The default Astro project structure includes a collection of folders and files that you can use to run and customize Airflow. For this tutorial, you only need to know the following files and folders:

> - `/dags`: A directory of DAG files. Each Astro project includes one example DAG `example_astronauts`. For more information on DAGs, see [Introduction to Airflow DAGs](dags.md).

Reviewer: ```suggestion
- `/dags`: A directory of DAG files. Each Astro project includes one example DAG: `example_astronauts`. For more information on DAGs, see [Introduction to Airflow DAGs](dags.md).
```

![Screenshot of the DAGs page of the Airflow UI showing the example_astronauts DAG with no run history yet.](/img/tutorials/get-started-with-airflow_ui_with_starter_dags.png)

Each DAG is listed with a few of its properties, including tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!

## Step 4: Trigger a DAG run

The `example_astronauts` DAG that was generated with your Astro project is a simple ETL pipeline with two tasks:

- `get_astronauts`: queries the [Open Notify API](https://github.com/open-notify/Open-Notify-API) for information about astronauts currently in space. The task returns the list of dictionaries containing the name and the spacecraft of all astronauts currently in space, which is passed to the second task in the DAG. This tutorial does not go into depth on how to pass data between tasks, but you can learn more about it in the [Pass data between tasks](airflow-passing-data-between-tasks.md) guide.

> - `print_astronaut_craft`: is a dynamically mapped Airflow task. Creating one dynamically mapped task instance for each astronaut in space, which each prints a statement to the logs for one astronaut. Dynamic task mapping is a versatile feature of Airflow that allows you to create a variable number of tasks at runtime. This feature is covered in more depth in the [Create dynamic Airflow tasks](dynamic-tasks.md) guide.

Reviewer: ```suggestion
- `print_astronaut_craft`: is a dynamically mapped Airflow task. It creates one dynamically mapped task instance for each astronaut in space, which prints a statement to the logs for that astronaut. Dynamic task mapping is a versatile feature of Airflow that allows you to create a variable number of tasks at runtime. This feature is covered in more depth in the [Create dynamic Airflow tasks](dynamic-tasks.md) guide.
```

Let's explore the available views in the **DAGs** page. To access different DAG views for `example_astronauts`:

1. Click the name of the DAG to access the default **Grid** view, which shows the state of completed and currently running tasks.

    Each column in the grid represents a complete DAG run, and each block in the column represents a specific task instance. This view is useful for seeing DAG runs over time and troubleshooting previously failed task instances.

    ![Screenshot of the Airflow Grid view showing two DAG runs of the example_astronauts DAG.](/img/tutorials/get-started-with-airflow_grid_view.png)

    Click on a green square to display additional information about the related task instance on the right side of the Airflow UI. The task instance view includes tabs with additional information for the task instance, such as its logs and historic runs. This is one of many available views that show details about your DAG.

    ![Gif showing how to access the task logs of a regular task instance by clicking on its task instance square in the Grid view and then on the Logs tab.](/img/tutorials/get-started-with-airflow_access_task_instance.gif)

>     To access information about [dynamically mapped](dynamic-tasks.md) task instances of a a dynamically mapped task, click the green square of the task instance and then click on **[] Mapped task** to get a list of all dynamically mapped task instances. Click on any entry in the list to access information about the dynamically mapped task instance, such as the logs.

Reviewer 2: I thought "mapping task" / "mapped task" might be clear clear, although you could also say the mapping task was "mapped" to each instance. Hmmmm need to think about this more

## Step 3: Log in to the Airflow UI

The [Airflow UI](airflow-ui.md) is essential for managing Airflow. It contains information about your current DAG Runs and is the best place to create and update Airflow connections to third-party data services.

To access the Airflow UI, open `http://localhost:8080/` in a browser and log in with `admin` for both your username and password.

The default page in the Airflow UI is the **DAGs** page, which shows an overview of all DAGs in your Airflow environment:

![View of starter DAG in the Airflow UI](/img/tutorials/get-started-with-airflow_ui_with_starter_dags.png)

Each DAG is listed with a few of its properties, including tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!

## Step 4: Trigger a DAG run

A **DAG run** is an instance of a DAG running on a specific date. Let's trigger a run of the `example_astronauts` DAG that was generated with your Astro project.

> To provide a basic demonstration of an ETL pipeline, this DAG queries the [Open Notify API](https://github.com/open-notify/Open-Notify-API) to get a list of all astronauts currently in space, passes that list to a second task and dynamically prints a statement to the Airflow logs for each astronaut. 

Reviewer: ```suggestion
To provide a basic demonstration of an ETL pipeline, this DAG queries the [Open Notify API](https://github.com/open-notify/Open-Notify-API) to get a list of all astronauts currently in space, passes that list to a second task, and dynamically prints a statement to the Airflow logs for each astronaut. 
```

Let's explore the available views in the **DAGs** page. To access different DAG views for `example_astronauts`:

1. Click the name of the DAG.

    The default DAG view is the **Grid** view, which shows the state of completed and currently running tasks. Each column in the grid represents a complete DAG run, and each block in the column represents a specific task instance. This view is useful for seeing DAG runs over time and troubleshooting previously failed task instances.

    ![Screenshot of the Airflow Grid view showing two DAG runs of the example_astronauts DAG.](/img/tutorials/get-started-with-airflow_grid_view.png)

    Click on a green square to display additional information about the related task instance on the right side of the Airflow UI. The task instance view includes tabs with additional information for the task instance, such as its logs and historic runs. This is one of many available views that show details about your DAG.

    ![Access task logs](/img/tutorials/get-started-with-airflow_access_task_instance.gif)

>     To access information about dynamically mapped task instances of a a dynamically mapped task, click the green square of the task instance and then click on **[] Mapped task** to get a list of all dynamically mapped task instances. Click on any entry in the list to access information about the dynamically mapped task instance, such as the logs.

Reviewer: I think we should add a note here that this tutorial doesn't discuss dynamic task mapping in depth, but it is a frequently used feature and you can find more info at [link to the doc], etc. Otherwise it seems a bit out of place to mention the feature but not define it.

  :::

> ## Step 6: Make a change to the example DAG

Reviewer: Honestly I'm not sure this step really adds anything educationally - can we remove it and go straight to writing a new DAG?

Now that you can run DAGs and navigate the UI, let's make a change to the `example_astronauts` DAG and run it again.

1. Open the `example_astronauts.py` file in your IDE. This file is located in the `/dags` folder of your Astro project. This DAG is a simple example of an ETL pipeline with two tasks: 

>     - The first task, `get_astronauts`, queries the [Open Notify API](https://github.com/open-notify/Open-Notify-API) for information about astronauts currently in space. The task returns the list of dictionaries containing the name and the spacecraft of all astronauts currently in space, which is passed to the second task in the DAG. This tutorial does not go into depth on how to pass data between tasks, but you can learn more about it in the [Pass data between tasks](airflow-passing-data-between-tasks.md) guide.

Reviewer: I think this description should go in Step 4 when we first introduce the DAG. Here it seems a little repetitive, and odd that we don't mention that earlier when looking at all the views of the DAG.

    ```

## Step 7: Write a new DAG

Now that we can run DAGs and navigate the UI, let's write our own DAG and run it. 

In this step, you'll write a DAG that:

- Retrieves the number of people currently in space from the Airflow XCom table and prints it to the logs. This table is part of the Airflow metadata database and is used to pass messages between tasks. The `example_astronauts` DAG already pushed the number of astronauts to XCom, when you run it in [Step 4](#step-4-trigger-a-dag-run).
- Runs a bash statement reacting to the number of people in space.

You'll copy most of the code, trigger the DAG, and then confirm the expected output is returned.

1. Create a new Python file in the `/dags` folder named `my_astronauts_dag.py`.
2. Open `my_astronauts_dag.py` in your IDE. Add the docstring explaining what this DAG will do at the top of the file:

    ```python 
    """
    ## Print the number of people currently in space

    This DAG will pull the number of people currently in space. The number is pulled
    from XCom and was pushed by the `get_astronauts` task in the `example_astronauts` DAG.
    """
    ```

>     Adding a docstring to your DAG is a best practice. It helps you and others understand what the DAG does and how it works and it can be converted to a [DAG Doc] by setting `doc_md=__doc__` in the DAG definition. 

Reviewer: ```suggestion
    Adding a docstring to your DAG is a best practice. It helps you and others understand what the DAG does and how it works and it can be converted to a [DAG Doc](custom-airflow-ui-docs-tutorial.md) by setting `doc_md=__doc__` in the DAG definition. 
```
Double check me on the link here :)

        """

        num_people_in_space = context["ti"].xcom_pull(
            dag_id="example_astronauts",
            task_ids="get_astronauts",
            key="number_of_people_in_space",
            include_prior_dates=True,
        )

        print(f"There are currently {num_people_in_space} people in space.")
    ```

    The `@task` decorator is used to define a task in the TaskFlow API. By default, the `task_id` will be the name of the Python function. Each task in an Airflow DAG requires a unique `task_id`. This `print_num_people_in_space` task pulls the number of people in space from XCom and prints it to the logs. Learn more about XCom in the [Pass data between tasks](airflow-passing-data-between-tasks.md) guide.

    You can put this decorator on top of any Python function to turn your code into an Airflow task. This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.

6. Add a second task to your DAG:

    ```python
    print_reaction = BashOperator(
        task_id="print_reaction",
        bash_command="echo This is awesome!",
    )
    ```

>     An operator is a Python class containing the logic to define the work to be completed by a single task. This second task `print_reaction` uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/bashoperator) to run a bash command that prints `This is awesome!` to the logs. The first parameter of the task (`task_id`) defines the name of the task that appears in the Airflow UI. 

Reviewer: ```suggestion
    For common tasks in Airflow, you can use pre-built operators instead of writing your own function. An operator is a Python class containing the logic to define the work to be completed by a single task. This second task, `print_reaction`, uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/bashoperator) to run a bash command that prints `This is awesome!` to the logs. The first parameter of the task (`task_id`) defines the name of the task that appears in the Airflow UI. 
```
Link to operators guide here?

7. Define the dependencies between the two tasks using the `chain` function:

    ```python
    chain(print_num_people_in_space(), print_reaction)
    ```

    The `chain` function is used to define task dependencies. In this example, `print_num_people_in_space` is **upstream** of `print_reaction`, meaning that `print_num_people_in_space` must finish before `print_reaction` can start. For more information on how to define task dependencies, see [Managing Dependencies in Apache Airflow](managing-dependencies.md).

    Alternatively you can use bitshift operators to define the dependencies between the two tasks:

    ```python
    print_num_people_in_space >> print_reaction
    ```

    This notation is unique to Airflow and is functionally equivalent to the `chain` function shown above.

8. Save your code.

## Step 8: Run the new DAG

> To view your new DAG in the Airflow UI, enter `http://localhost:8080/` in your browser. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files are parsed for changes every 30 seconds, while new files can be detected every 5 minutes.

Reviewer: ```suggestion
Go back to the Airflow UI to view your new DAG. Airflow will parse the `/dags` directory for changes to existing files  every 30 seconds and new files every 5 minutes.
```
They should already have the Airflow UI up at this point in the tutorial, so I think we can simplify.

    - The `datetime` class from the [pendulum package](https://pendulum.eustace.io/), which is used to define the `start_date` of the DAG.

4. Instantiate a DAG using the `@dag` decorator on top of a Python function:

    ```python
    @dag(
        dag_id="my_astronauts_dag",
        start_date=datetime(2024, 1, 1),
        schedule=[Dataset("current_astronauts")],
        catchup=False,
        doc_md=__doc__,
        default_args={"owner": "Astro", "retries": 3},
        tags=["My First DAG!"],
    )
    def my_astronauts_dag():
    ```

    `@dag` instantiates a DAG context in which tasks can be defined and given dependencies. The instantiation includes several important arguments:

    - `dag_id`: The name of the DAG that appears in the Airflow UI. If no `dag_id` is specified, the name of the Python function is used as the DAG ID. Each DAG must have a unique name, and Astronomer recommends using the same name for the DAG file and the `dag_id`.
    - `start_date`: The date and time when the DAG is scheduled to start running, given as a datetime object. In this example, the DAG is triggered on its schedule as long as the current time is 0:00 UTC on January 1st, 2024 or later.

>     - `schedule`: The frequency the DAG runs. There are many options to define the schedule of a DAG, including as a [timedelta object](https://docs.python.org/3/library/datetime.html#timedelta-objects), a [CRON expression](https://crontab.guru/), or as a macro such as `@daily`. If you don't set this value, the DAG runs every 24 hours after the `start_date`. In this example, the DAG is scheduled to run whenever the `example_astronauts` DAG's `get_astronauts` task completes successfully using an Airflow Dataset, for more information see [Datasets and data-aware scheduling in Airflow](airflow-datasets.md).

Reviewer 2: ```suggestion
    - `schedule`: The frequency at which the DAG runs. If you don't set this value, the DAG runs every 24 hours after the `start_date`. In this example, the DAG is scheduled to run whenever the `example_astronauts` DAG's `get_astronauts` task completes successfully using an Airflow dataset. There are many options to define the schedule of a DAG. For more information, see [DAG scheduling and time tables in Airflow](https://docs.astronomer.io/learn/scheduling-in-airflow).
```
Simplified some of the linking here to focus on "schedule" as a concept rather than datasets

    - `default_args`: A dictionary of arguments to be passed to every task in the DAG. In this example, the `default_args` change the owner of the DAG and set the default number of retries in case of a failure for all tasks in the DAG to 3. You can pass any arguments to all your tasks by setting them in `default_args` and override them for specific tasks by setting them in the task definition.
    - `tags`: Defines the **Tags** that appear in the **DAGs** page of the Airflow UI. These can help you organize DAGs in more complex projects.

5. Add a task to your DAG:

    ```python
    @task
    def print_num_people_in_space(**context) -> None:
        """
        This task pulls the number of people currently in space from XCom. The number is
        pushed by the `get_astronauts` task in the `example_astronauts` DAG.
        """

        num_people_in_space = context["ti"].xcom_pull(
            dag_id="example_astronauts",
            task_ids="get_astronauts",
            key="number_of_people_in_space",
            include_prior_dates=True,
        )

        print(f"There are currently {num_people_in_space} people in space.")
    ```

>     The `@task` decorator is used to define a task in the TaskFlow API. By default, the `task_id` is the name of the Python function. Each task in an Airflow DAG requires a unique `task_id`. This `print_num_people_in_space` task pulls the number of people in space from XCom and prints it to the logs. Learn more about XCom in the [Pass data between tasks](airflow-passing-data-between-tasks.md) guide.

Reviewer 2: ```suggestion
    The `@task` decorator lets you define Python functions with very little boilerplate code. This is the syntax of the TaskFlow API, which is Airflow's lightweight framework for defining tasks. This `print_num_people_in_space` task pulls the number of people in space from XCom and prints it to the logs. Learn more about XCom in the [Pass data between tasks](airflow-passing-data-between-tasks.md) guide.
```
This is the first proper introduction of the concept of the TaskFlow API, so we should add a half sentence about why it exists.

>     You can put this decorator on top of any Python function to turn your code into an Airflow task. This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.

Reviewer 2: ```suggestion
    You can put this decorator on top of any Python function to turn your code into an Airflow task. This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.
    
    Each task in an Airflow DAG requires a unique `task_id`. Here we can see why that's important: the `num_people_in_space` task can call the unique `get_astronauts` task to pull the data it generated from XCom.
```

6. Add a second task to your DAG:

    ```python
    print_reaction = BashOperator(
        task_id="print_reaction",
        bash_command="echo This is awesome!",
    )
    ```

>     For common tasks in Airflow, you can use pre-built [operators](what-is-an-operator.md) instead of writing your own function. An operator is a Python class containing the logic to define the work to be completed by a single task. This second task, `print_reaction`, uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/bashoperator) to run a bash command that prints `This is awesome!` to the logs. The first parameter of the task (`task_id`) defines the name of the task that appears in the Airflow UI. 

Reviewer 2: ```suggestion
    For common operations, you can use pre-built [operators](what-is-an-operator.md) instead of writing your own function. An operator is a Python class containing the logic to define the work to be completed by a single task. `print_reaction` uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/bashoperator) to run a bash command.
```
We mention task ID earlier, so cutting this explanation

7. Define the dependencies between the two tasks using the `chain` function:

    ```python
    chain(print_num_people_in_space(), print_reaction)
    ```

    The `chain` function is used to define task dependencies. In this example, `print_num_people_in_space` is **upstream** of `print_reaction`, meaning that `print_num_people_in_space` must finish before `print_reaction` can start. For more information on how to define task dependencies, see [Managing Dependencies in Apache Airflow](managing-dependencies.md).

    Alternatively you can use bitshift operators to define the dependencies between the two tasks:

    ```python
    print_num_people_in_space >> print_reaction
    ```

    This notation is unique to Airflow and is functionally equivalent to the `chain` function shown above.

> 8. Save your code.

Reviewer 2: We do have this ability! See: https://docusaurus.io/docs/markdown-features#details . It has to be in a "details" block, but that might be nice as a way to label the content while it's still condensed.

> 8. Save your code.

Reviewer 2: Let me know if you'd like me to implement it, tbh I think having the full project code will be helpful for the 2nd tutorial

> 8. Save your code.

Author: ooh nice! I made an attempt at implementing it and yeah can add it to the 2nd tutorial as well :)

## Step 7: Run the new DAG

Go back to the Airflow UI to view your new DAG. Airflow parses the `/dags` directory for changes to existing files every 30 seconds and new files every 5 minutes.

:::tip

You can manually trigger a full parse of your DAGs by running the following command in your terminal:

```sh
astro dev run dags reserialize
```

:::

When your new DAG appears in the Airflow UI, you can run it to test it.

1. Start the new DAG and trigger a run like you did in [Step 4](#step-4-trigger-a-dag-run).
2. Click the name of your new DAG and open the **Grid** view. After your DAG runs, there should be a green bar representing a successful run of the DAG.

    ![Screenshot of the Airflow Grid view showing a successful run of the my_astronauts_dag DAG.](/img/tutorials/get-started-with-airflow_grid_view_my_astronauts_dag.png)

> 3. The `my_astronauts_dag` is scheduled to run whenever the `current_astronauts` [Dataset](airflow-datasets.md) is updated by a successful run of the `get_astronauts` task in the `example_astronauts` DAG. Trigger another manual run of the `example_astronauts` DAG to see the `my_astronauts_dag` run again, as soon as the `get_astronauts` task has completed.

Reviewer 2: ```suggestion
3. The `my_astronauts_dag` is scheduled to run whenever the `current_astronauts` dataset is updated by a successful run of the `get_astronauts` task in the `example_astronauts` DAG. Trigger another manual run of the `example_astronauts` DAG to see the `my_astronauts_dag` run again after the `get_astronauts` task completes.
```
If there's a doc link we want to link to many times, it might be worth including in a "See also" section a the bottom of the doc. There's a lot of links in the doc at the moment, and I worry that new users might get overwhelmed if they start clicking around into new guides.

## Step 8: View task logs

> When you tell Airflow to print something to the terminal, the output appears in Airflow task logs. Task logs are an important feature for troubleshooting DAGs. If a task in your DAG fails, task logs are the best place to investigate why.

Reviewer: ```suggestion
When one of your tasks prints something, the output appears in Airflow task logs. Task logs are an important feature for troubleshooting DAGs. If a task in your DAG fails, task logs are the best place to investigate why.
```

7. Define the dependencies between the two tasks using the `chain` function:

    ```python
        chain(print_num_people_in_space(), print_reaction)
    ```

    The `chain` function is used to define task dependencies. In this example, `print_num_people_in_space` is **upstream** of `print_reaction`, meaning that `print_num_people_in_space` must finish before `print_reaction` can start. For more information on how to define task dependencies, see [Managing Dependencies in Apache Airflow](managing-dependencies.md).

    Alternatively you can use bitshift operators to define the dependencies between the two tasks:

    ```python
        print_num_people_in_space >> print_reaction
    ```

    This notation is unique to Airflow and is functionally equivalent to the `chain` function shown above.

8. Lastly, call the DAG function:

    ```python
    my_astronauts_dag()
    ```

> 9. Save your code. Your DAG should look like this:

Reviewer 2: ```suggestion
9. Save your code. Your completed DAG should look like this:
```
